### PR TITLE
Fix go example

### LINKED ISSUE #39: Update gocounter example
1. I think I removed the bpf code that this used :sweat_smile:  It might make more sense to use whatever cilium/ebpf recommends or just write the prog in bpf asm directly as it's only small. Was inspired by [this program](https://github.com/xdp-project/xdp-tutorial/tree/master/basic03-map-counter)
2. It now needs a certificate generated - `./scripts/certificates.sh` should be updated to allow easy generation of a client cert.

### PR DESCRIPTION
**Pull request by astoycos**
fixes #39 

This PR updates the gocounter example. Mainly it 
- Moves the bpf program's source code to this repo
- Integrates it with `cilium/ebpf`'s buit in build structure for compiling with `go generate` 
- Updates the bpf program to count bytes and packet count 
- Updates the go client to use mTLS 
- Updates the `certificates.sh` script to generate a cert explicitly for the go-counter example 

Also it fixes a bug where the xdp dispatcher program wasn't actually calling any attached programs due to a one off error (I can move that commit into it's own PR if desired EDIT::: Moved to -> https://github.com/redhat-et/bpfd/pull/54 This needs to merge first.

### CODE REVIEW

File: docs/developer/go.md [hunk 1]
```diff
@@ -1,3 +1,66 @@
...
+7 packets received
+1275 bytes received
+
+7 packets received
+1275 bytes received
+
+^CExiting...
+```
+
+Finally, press `ctl+c` when finished.
```
Comment (dave-tucker) [inline docs/developer/go.md:66]: s/ctl/ctrl/g

File: docs/developer/go.md [hunk 2]
```diff
@@ -1,3 +1,66 @@
...
+To build the userspace go client run:
+    
+    pushd examples/gocounter && go build && popd
+
+## Running 
+
+First start or ensure `bpfd` is up and running. 
+
+Then start the go program with:
+    pushd examples/gocounter && sudo ./gocounter <INTERNET INTERFACE NAME> && popd
```
Comment (dave-tucker) [inline docs/developer/go.md:40]: Nit. Should use backticks to make this a code block

File: examples/gocounter/main.go [hunk 3]
```diff
@@ -20,31 +21,46 @@ import (
...
 )
 
 type Stats struct {
 	Packets uint64
+	Bytes   uint64
 }
 
+const (
+	DefaultRootCaPath     = "/etc/bpfd/certs/ca/ca.pem"
+	DefaultClientCertPath = "/etc/bpfd/certs/go-bpfd-client/go-bpfd-client.pem"
```
Comment (dave-tucker) [inline examples/gocounter/main.go:34]: This should be gocounter.pem.
The model we are after is one certificate per application.

File: examples/gocounter/main.go [hunk 4]
```diff
@@ -20,31 +21,46 @@ import (
...
+	DefaultClientKeyPath  = "/etc/bpfd/certs/go-bpfd-client/go-bpfd-client.key"
+)
+
+// TODO (astoycos): Add back -Wall -Werror here once BTF is fixed in Aya
```
Comment (dave-tucker) [inline examples/gocounter/main.go:38]: Nit: When BTF maps are supported in Aya

File: scripts/certificates.sh [hunk 5]
```diff
@@ -14,6 +14,7 @@ init() {
     fi
     client bpfd
     client bpfctl
+    client go-bpfd-client
```
Comment (dave-tucker) [inline scripts/certificates.sh:17]: gocounter

File: docs/developer/go.md [hunk 6]
```diff
@@ -1,3 +1,66 @@
...
+
+The output should show the count and total bytes of packets as they pass through the
+interface as shown below:
+
+```
+[astoycos@localhost gocounter]$ sudo ./gocounter docker0
```
Comment (dave-tucker) [inline docs/developer/go.md:46]: You can probably strip everything before $

**Review by dave-tucker — CHANGES_REQUESTED** [review]
A few nits but this looks great! Thanks @astoycos

File: scripts/certificates.sh (hunk 5 above)
Comment (dave-tucker) [inline scripts/certificates.sh:17]: Maybe also in the docs you can call out how to use `./scripts/certificates.sh client gocounter` to generate this.
Or to say that a user needs to generate their own cerficate for their app, signed by the same CA as the bpfd certificate.

File: docs/developer/go.md (hunk 2 above)
Comment (astoycos) [reply docs/developer/go.md]: Done

**Comment by astoycos** [comment]
I think this is ready to go @dave-tucker just PTAL at the small commit 08321e8294e0154412410869e5f116d4a8445244 I added after the rebase to your changes from earlier today.

**Comment by Billy99** [comment]
@astoycos You still need part of #59. In [`bpfd/src/main.rs`, arou](https://github.com/redhat-et/bpfd/blob/main/bpfd/src/main.rs#L27), remove the `#[cfg(debug_assertions)]` and `#[cfg(not(debug_assertions))]` and set to "../../target/bpfel-unknown-none/release/xdp_dispatcher.bpf.o"

**Comment by astoycos** [comment]
Done, thanks @Billy99!

**Review by dave-tucker — APPROVED** [review]

**Pull request merged**